mmv1/third_party/terraform/services/resourcemanager/ephemeral_google_service_account_id_token.go

```diff
@@ -0,0 +1,251 @@
+	p.providerConfig = pd
+}
+
+func (p *googleEphemeralServiceAccountIdToken) Open(ctx context.Context, req ephemeral.OpenRequest, resp *ephemeral.OpenResponse) {
+	var data ephemeralServiceAccountIdTokenModel
+
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+
+	targetAudience := data.TargetAudience.ValueString()
+	creds := fwtransport.GetCredentials(ctx, fwmodels.ProviderModel{}, false, &resp.Diagnostics)
+
+	targetServiceAccount := data.TargetServiceAccount.ValueString()
+	// If a target service account is provided, use the API to generate the idToken
+	if targetServiceAccount != "" {
+		service := p.providerConfig.NewIamCredentialsClient(p.providerConfig.UserAgent)
+		name := fmt.Sprintf("projects/-/serviceAccounts/%s", targetServiceAccount)
+		DelegatesSetValue, _ := data.Delegates.ToSetValue(ctx)
+		tokenRequest := &iamcredentials.GenerateIdTokenRequest{
+			Audience:     targetAudience,
+			IncludeEmail: data.IncludeEmail.ValueBool(),
+			Delegates:    StringSet(DelegatesSetValue),
+		}
+		at, err := service.Projects.ServiceAccounts.GenerateIdToken(name, tokenRequest).Do()
+		if err != nil {
+			resp.Diagnostics.AddError(
+				"Error calling iamcredentials.GenerateIdToken",
+				err.Error(),
+			)
+			return
+		}
+
+		data.IdToken = types.StringValue(at.Token)
+		resp.Diagnostics.Append(resp.Result.Set(ctx, data)...)
+		return
+	}
+
+	// If no target service account, use the default credentials
+	ctx = context.Background()
+	co := []option.ClientOption{}
+	if creds.JSON != nil {
+		co = append(co, idtoken.WithCredentialsJSON(creds.JSON))
+	}
+
+	idTokenSource, err := idtoken.NewTokenSource(ctx, targetAudience, co...)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Unable to retrieve TokenSource",
+			err.Error(),
+		)
+		return
+	}
+	idToken, err := idTokenSource.Token()
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Unable to retrieve Token",
+			err.Error(),
+		)
+		return
+	}
+
+	data.IdToken = types.StringValue(idToken.AccessToken)
+	resp.Diagnostics.Append(resp.Result.Set(ctx, data)...)
+}
+
+func StringSet(d basetypes.SetValue) []string {
+
+	StringSlice := make([]string, 0)
+	for _, v := range d.Elements() {
+		StringSlice = append(StringSlice, v.(basetypes.StringValue).ValueString())
+	}
+	return StringSlice
+}
+
+var serviceAccountNamePatterns = []string{
+	`^.+@.+\.iam\.gserviceaccount\.com$`,                     // Standard IAM service account
+	`^.+@developer\.gserviceaccount\.com$`,                   // Legacy developer service account
+	`^.+@appspot\.gserviceaccount\.com$`,                     // App Engine service account
+	`^.+@cloudservices\.gserviceaccount\.com$`,               // Google Cloud services service account
+	`^.+@cloudbuild\.gserviceaccount\.com$`,                  // Cloud Build service account
+	`^service-[0-9]+@.+-compute\.iam\.gserviceaccount\.com$`, // Compute Engine service account
+}
+
+// Create a custom validator for service account names
+type serviceAccountNameValidator struct{}
```

Comment:
Please add some automated testing for this new validator logic

Reply:
added here https://github.com/GoogleCloudPlatform/magic-modules/pull/12268/files#diff-124849c72df16d146a9d4e8048f0c6fef6425d191b37eb58fef375fee23ffecdR70-R141

Since we intend on removing these once `fwvalidators` is merged. we can leave as is and then remove once its merged.